src/nbl/asset/utils/CSPIRVIntrospector.cpp

```diff
@@ -13,472 +13,528 @@ namespace nbl::asset
+            descriptorsToMerge[i].insert(descInfo);
         }
+        std::cout << debug.str() << std::endl;
     }
 
-    core::vector<SPushConstantRange> merged;
-
-    SPushConstantRange rngToPush; rngToPush.offset = 0u; rngToPush.size = 0u;
-    uint32_t stageFlags = 0u;
-    for (uint32_t i = 0u; i < ICPUMeshBuffer::MAX_PUSH_CONSTANT_BYTESIZE/sizeof(uint32_t); i += sizeof(uint32_t))
+    // validate if only descriptors with the highest bindings are run-time sized
+    std::array<int32_t, ICPUPipelineLayout::DESCRIPTOR_SET_COUNT> highestBindingsTmp = m_highestBindingNumbers;
+    for (uint32_t i = 0u; i < ICPUPipelineLayout::DESCRIPTOR_SET_COUNT; ++i)
     {
-        SPushConstantRange curr; curr.offset = i; curr.size = sizeof(uint32_t);
-
-        uint32_t tmpFlags = 0u;
-        for (uint32_t stg=0u; stg<infos.size(); stg++)
-            for (const SPushConstantRange& rng : ranges[stg])
-                if (curr.overlap(rng))
-                    tmpFlags |= rng.stageFlags;
+        const auto& introBindingInfos = stageData->getDescriptorSetInfo(i);
 
-        if (!i)
-            stageFlags = tmpFlags;
+        for (const auto& stageIntroBindingInfo : introBindingInfos)
+            highestBindingsTmp[i] = std::max<const int32_t>(highestBindingsTmp[i], stageIntroBindingInfo.binding);
 
-        if (!tmpFlags)
-            continue;
-        if (tmpFlags == stageFlags)
-            rngToPush.size += sizeof(uint32_t);
-        else
+        for (const auto& descriptor : descriptorsToMerge[i])
         {
-            rngToPush.stageFlags = static_cast<ICPUShader::E_SHADER_STAGE>(stageFlags);
-            merged.push_back(rngToPush);
-            stageFlags = 0u;
-
-            rngToPush.offset = i;
-            rngToPush.size = sizeof(uint32_t);
+            if (descriptor.binding < highestBindingsTmp[i] && descriptor.isRuntimeSized())
+                return false;
         }
     }
-    if (stageFlags)
-    {
-        rngToPush.stageFlags = static_cast<ICPUShader::E_SHADER_STAGE>(stageFlags);
-        merged.push_back(rngToPush);
-    }
 
-    core::smart_refctd_dynamic_array<SPushConstantRange> rngsArray;
-    if (merged.size())
+    // validation successfull, update `CPipelineIntrospectionData` contents
+    m_highestBindingNumbers = highestBindingsTmp;
+    for (uint32_t i = 0u; i < ICPUPipelineLayout::DESCRIPTOR_SET_COUNT; ++i)
+        m_descriptorSetBindings[i].merge(descriptorsToMerge[i]);
+
+    // TODO: possible push constants validation
+    // can only be success now
+    const auto& pc = stageData->getPushConstants();
+    auto a = pc.size;
+    if (pc.present())
     {
-        rngsArray = core::make_refctd_dynamic_array<decltype(rngsArray)>(merged.size());
-        memcpy(rngsArray->data(),merged.data(),rngsArray->size()*sizeof(SPushConstantRange));
+        std::span<core::bitflag<ICPUShader::E_SHADER_STAGE>> pcRangesSpan(
+            &m_pushConstantBytes[pc.offset],
+            &m_pushConstantBytes[pc.offset + pc.size]
```

Comment:
heads up, with a `T data[256]` doing `&data[256]` is bad practice because `operator[]` and `&` might actually perform some operation (because of overrides) on an out-of-bound memory location, and address sanitizers might moan

the correct syntax here is `pcRangesSpan = {m_pushCosntantBytes+pc.offset,pc.size}`